lib/steep/type_construction.rb

```diff
@@ -2401,7 +2401,17 @@ def synthesize(node, hint: nil, condition: false)
             if value_node
               type, constr = synthesize(value_node, hint: hint)
 
-              if hint.is_a?(AST::Types::Proc) && value_node.type == :sym
+              if hint.is_a?(AST::Types::Proc) && value_node.type == :send && value_node.children[1] == :method && type.is_a?(AST::Types::Name::Instance) && type.name.to_s == '::Method'
+                method_name = value_node.children[2].children[0]
+                if method = calculate_interface(AST::Types::Self.instance, private: true)&.methods&.[](method_name)
+                  # TODO: select appropriate method type from `method.method_types`.
```

Comment:
I'm unsure how to choose the appropriate method type from `method.method_types`. So I use the first one.